pkg/debuginfo/debuginfod.go

```diff
@@ -0,0 +1,44 @@
+package debuginfo
+
+import (
+	"io"
+	"net/http"
+	"path"
+)
+
+var publicServer = "https://debuginfod.systemtap.org"
+
+type DebugInfodClient interface {
+	GetDebugInfo(buildid string) (io.ReadCloser, error)
+}
+
+type HttpDebuginfodClient struct {
+	UpstreamServer string //url
+}
+type ObjectStorageDebugInfodClientCache struct {
+	client DebugInfodClient
+}
+
+func NewHttpDebugInfoClient(serverUrl string) HttpDebuginfodClient {
+	return HttpDebuginfodClient{UpstreamServer: serverUrl}
+}
+
+func NewObjectStorageDebugInfodClientCache(h HttpDebuginfodClient) *ObjectStorageDebugInfodClientCache {
+	return &ObjectStorageDebugInfodClientCache{client: &h}
+}
+
+func (c *ObjectStorageDebugInfodClientCache) GetDebugInfo(buildid string) (io.ReadCloser, error) {
+	return c.client.GetDebugInfo(buildid)
+}
+
+func (c *HttpDebuginfodClient) GetDebugInfo(buildID string) (io.ReadCloser, error) {
+	serverUrl := path.Join(c.UpstreamServer, buildID, "debuginfo")
```

Comment:
This is why we want to parse the upstream server URL when we create the client, as `path.Join` should only be called on the actual path, instead of the whole URL.